File: 0148. Sort List/Solution.java
//using Merge Sort

class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null || head.next==null){
            return head;
        }

        ListNode mid = findMid(head);

        ListNode left = head;
        ListNode right = mid.next;

        mid.next = null;

        left = sortList(left);
        right = sortList(right);

        return mergeSorted(left,right);
    }

    public ListNode mergeSorted(ListNode head1,ListNode head2){
        ListNode h1 = head1;
        ListNode h2 = head2;

        ListNode res = new ListNode(0);
        ListNode r = res;

        while(h1!=null && h2!=null){
            if(h1.val<h2.val){
                r.next = h1;
                r = r.next;
                h1 = h1.next;
            }
            else {
                r.next = h2;
                r = r.next;
                h2 = h2.next;
            }
        }

        while(h1!=null){
            r.next = h1;
            r = r.next;
            h1 = h1.next;
        }

        while(h2!=null){
            r.next = h2;
            r = r.next;
            h2 = h2.next;
        }
    return res.next;
    }

    public ListNode findMid(ListNode head){
        if(head==null)return null;
        ListNode fast = head.next;
        ListNode slow = head;

        while(fast!=null && fast.next!=null){
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
}

// using Bubble Sort

// class Solution {
//     public ListNode sortList(ListNode head) {
//         ListNode h1 = head;

//         while(h1!=null){
//             ListNode h2 = h1;
//             while(h2!=null){
//                 if(h1.val>h2.val){
//                     int temp = h1.val;
//                     h1.val = h2.val;
//                     h2.val = temp;
//                 }
//                 h2 = h2.next;
//             }
//             h1 = h1.next;
//         }

//         return head;
//     }
// }

File: 0148. Sort List/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def sortList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """

        if not head or not head.next:
            return head

        mid = self.findMid(head)

        left = head
        right = mid.next
        mid.next = None

        left = self.sortList(left)
        right = self.sortList(right)

        return self.mergeSorted(left,right)



    def findMid(self,head):
        if not head:
            return head
        fast = head.next
        slow = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        return slow

    def mergeSorted(self,head1,head2):
        res = ListNode(0)
        r = res

        p1 = head1
        p2 = head2

        while p1 and p2:
            if(p1.val<p2.val):
                r.next = p1
                r = r.next
                p1 = p1.next
            else:
                r.next = p2
                r = r.next
                p2 = p2.next

        while p1:
            r.next = p1
            r = r.next
            p1 = p1.next

        while p2:
            r.next = p2
            r = r.next
            p2 = p2.next

        return res.next

        